### ISSUE TITLE
iOS 16 may not respect a request to change orientation

### ISSUE THREAD

**Issue by msmtamburro** [apache/cordova-plugin-screen-orientation#100]
# Bug Report

## Problem

Getting this issue on our radar early:

https://developer.apple.com/forums/thread/707735?answerId=721941022#721941022

### What is expected to happen?

This approach used to work, but may not on iOS 16:
 [[UIDevice currentDevice] setValue:@(orientation) forKey:@"orientation"]

We may need to switch to:
https://developer.apple.com/documentation/uikit/uiwindowscene/3975944-requestgeometryupdate/
(which would require building from XCode 14)

### Version information

iOS 16 beta 4

## Checklist

- [ ] I searched for existing GitHub issues
- [ ] I updated all Cordova tooling to most recent version
- [ ] I included all the necessary information above

**Comment by msmtamburro** [comment]
Specific guidance:
https://developer.apple.com/documentation/ios-ipados-release-notes/ios-ipados-16-release-notes

[UIViewController shouldAutorotate] has been deprecated is no longer supported. [UIViewController attemptRotationToDeviceOrientation] has been deprecated and replaced with [UIViewController setNeedsUpdateOfSupportedInterfaceOrientations].

Workaround: Apps relying on shouldAutorotate should reflect their preferences using the view controllers supportedInterfaceOrientations. If the supported orientations change, use `-[UIViewController setNeedsUpdateOfSupportedInterface

**Comment by msmtamburro** [comment]
And a code snippet that will probably work (where `self` is your view controller):
```
    if (@available(iOS 16.0, *)) {
#if __IPHONE_OS_VERSION_MAX_ALLOWED > __IPHONE_15_5 // Xcode 14 and iOS 16, or greater
        [self setNeedsUpdateOfSupportedInterfaceOrientations];
#endif
    } else {
        [UIViewController attemptRotationToDeviceOrientation];
    }
```

**Comment by msmtamburro** [comment]
Replacing [line 78](https://github.com/apache/cordova-plugin-screen-orientation/blob/master/src/ios/CDVOrientation.m#L78) with something like this might work (untested):

```
                if (@available(iOS 16.0, *)) {
#if __IPHONE_OS_VERSION_MAX_ALLOWED > __IPHONE_15_5 // Xcode 14 and iOS 16, or greater
                    [self.viewController setNeedsUpdateOfSupportedInterfaceOrientations];
#endif
                } else {
                    [UINavigationController attemptRotationToDeviceOrientation];
                }
```

**Comment by tookiez** [comment]
Hello there!

i've tried replacing 

`[UINavigationController attemptRotationToDeviceOrientation]`

with your code. Unfortunately when i force the orientationChange, it doesn't trigger.
Meanwhile i cannot find any indication or clear example to use **requestGeometryUpdate**

**Comment by tory37** [comment]
Upvoting that suggest change did not work for us either.

**Comment by msmtamburro** [comment]
Apologies that I still haven't had time to play with this.  For diagnostics: When you're trying to use `setNeedsUpdateOfSupportedInterfaceOrientations` from Xcode 14 RC on iOS 16, does your list of supported orientations _only_ include the desired orientation?

**Comment by Robertndrei** [comment]
Hi, the fix is working, but you have to replace[ line 83](https://github.com/apache/cordova-plugin-screen-orientation/blob/master/src/ios/CDVOrientation.m#L83)
`[[UIDevice currentDevice] setValue:value forKey:@"orientation"];`

**Comment by tookiez** [comment]
@Robertndrei ok, i've replaced every

`[UIDevice currentDevice] setValue:value forKey:@"orientation"]`

and

`[UINavigationController attemptRotationToDeviceOrientation]`

with 

```
  if (@available(iOS 16.0, *)) {
      #if __IPHONE_OS_VERSION_MAX_ALLOWED > __IPHONE_15_5 // Xcode 14 and iOS 16, or greater
          [self.viewController setNeedsUpdateOfSupportedInterfaceOrientations];
      #endif
  } else {
      [UINavigationController attemptRotationToDeviceOrientation];
  } 
```

and now seems work correctly

**Comment by chrisvasz** [comment]
@tookiez's fix (replacing lines 78 and 83 with the [suggestion from above](https://github.com/apache/cordova-plugin-screen-orientation/issues/100#issuecomment-1212391656)) worked for me too. How can we help get this fix released?

**Comment by Robertndrei** [comment]
Actually only line 83 is necesarry to be replaced

**Comment by jansgescheit** [comment]
@jaydrogers i tested your fork on a Simulator with a iPhone X on iOS 12.4 Compiled on XCode 14. Add the moment screen orientation does not lock in any directions

**Comment by jaydrogers** [comment]
Strange!

If you find issues, chime in on this PR. Looks like a contributor is looking at it now: https://github.com/apache/cordova-plugin-screen-orientation/pull/102

**Comment by laughsky** [comment]
if use xcode 13，the following is better:
```
if (@available(iOS 16.0, *)) {
    #if __IPHONE_OS_VERSION_MAX_ALLOWED > __IPHONE_15_5
        SEL supportedInterfaceSelector = NSSelectorFromString(@"setNeedsUpdateOfSupportedInterfaceOrientations");
        [self.viewController performSelector:supportedInterfaceSelector];
    #endif
} else {
    [UINavigationController attemptRotationToDeviceOrientation];
}
```
